### DISCUSSION TITLE
`nvim_exec_lua` from visual mode?

### DISCUSSION THREAD

**Discussion by adigitoleo** [neovim/neovim#31720] [Q&A] [answered]
In a test suite for a plugin, I would like to use `nvim_exec_lua` (actually wrapped via `lua_get` of [mini.test](https://github.com/echasnovski/mini.nvim/blob/main/doc/mini-test.txt)) to run an API function. The behaviour of the function depends on the result of `vim.api.nvim_get_mode().mode`, but I find that even if I use `nvim_input` to set up a visual selection (e.g. `nvim_input('V2j')`), when I subsequently call the API function using `nvim_exec_lua` the `.mode` is always `'n'`.

In other words, it doesn't seem possible to use `nvim_exec_lua` from visual mode. I confirmed this by entering visual mode and trying to run `'<,'>=vim.api.nvim_exec_lua('print("wow")')`, which results in an E5108 error (i.e. an attempt to call `nil`; the `api` module doesn't have a `nvim_exec_lua` element in this case)

I guess my questions are: am I correct that it is impossible? if so, why (and why is it not documented)? and is there any known workaround?

Cheers

**Comment by adigitoleo** [comment]
Perhaps @echasnovski can also help as to the details of the `lua_get` implementation in the test framework.

**Comment by echasnovski** [comment, marked as answer]
The `nvim_exec_lua` itself is [RPC only](https://neovim.io/doc/user/api.html#nvim_exec_lua()), meaning that executing something like `:=vim.api.nvim_exec_lua('return 1+1')` will fail when called from the current (interactive) session. It will work only through RPC connection (like `child.api.nvim_exec_lua` from 'mini.test').

Both `child.lua` and `child.lua_get` from 'mini.test' are [wrappers around `nvim_exec_lua()`](https://github.com/echasnovski/mini.nvim/blob/2011aff270bcd3e1f3ad088253ace2d574967bed/doc/mini-test.txt#L778-L781).

So the answer is no, `nvim_exec_lua()` works in any mode, but only through RPC. For example, the following code works in 'mini.test' and `mode` is equal to `'V'`:

```lua
child.api.nvim_input('V2j')
local mode = child.api.nvim_exec_lua('return vim.api.nvim_get_mode().mode', {})
```
